runtime/hydradx/src/evm/evm_error_decoder.rs

```diff
@@ -0,0 +1,85 @@
+use crate::Liquidation;
+use codec::{Decode, DecodeLimit};
+use frame_support::traits::Get;
+use hydradx_traits::evm::CallResult;
+use pallet_evm::{ExitError, ExitReason};
+use sp_runtime::format;
+use sp_runtime::traits::Convert;
+use sp_runtime::DispatchError;
+use sp_std::boxed::Box;
+use sp_std::vec::Vec;
+
+const ERROR_STRING_SELECTOR: [u8; 4] = [0x08, 0xC3, 0x79, 0xA0]; // Error(string)
+const PANIC_SELECTOR: [u8; 4] = [0x4E, 0x48, 0x7B, 0x71]; // Panic(uint256)
+const FUNCTION_SELECTOR_LENGTH: usize = 4;
+pub const MAX_DECODE_DEPTH: u32 = 256; // Used for DispatchError decoding to prevent stack exhaustion attacks
+pub const MAX_ERROR_DATA_LENGTH: usize = 1024; // Maximum length of EVM error data to prevent DOS attacks
+
+pub struct EvmErrorDecoder;
+
+impl Convert<CallResult, DispatchError> for EvmErrorDecoder {
+	fn convert(call_result: CallResult) -> DispatchError {
+		if let ExitReason::Error(ExitError::OutOfGas) = call_result.exit_reason {
+			return pallet_dispatcher::Error::<crate::Runtime>::EvmOutOfGas.into();
+		}
+
+		// DOS Prevention: Limit error data size to prevent memory exhaustion attacks
+		if call_result.value.len() > MAX_ERROR_DATA_LENGTH {
+			log::warn!(
+				target: "evm::error_decoder",
+				"EVM error data too large: {} bytes (max: {}). Truncating to prevent DOS.",
+				call_result.value.len(),
+				MAX_ERROR_DATA_LENGTH
+			);
+			let truncated_value = call_result.value[..MAX_ERROR_DATA_LENGTH].to_vec();
```

Comment:
you can improve this. it is unnecessary vec allocation.

you can use slice instead, and there wont be any copy involved.